descs: rewrite Collection backend

postamar opened the pull request:
This rewrite was motivated by the upcoming need to validate descriptors
up to different levels depending on why they were being read.

This requires tracking the validation level in a more granular fashion
in the Collection's caches. This in turn required dissociating the
uncommitted descriptors from the stored descriptors again, to allow the
stored descriptors cache to more closely mirror what's physically being
stored in the system.descriptor table.

Validation is now performed lazily and all descriptor queries now share
a common structure, be it by-ID lookups, by-name lookups, or full table
scans: cache lookups, reads, validation and hydration all happen in the
same sequence.

Despite significant rewriting of the inner workings of the descriptors
collection, this commit should not functionally change anything.

Informs #85263.

Release justification: low-risk, high-benefit change.
Release note: None

Commit:
descs: rewrite Collection backend

This rewrite was motivated by the upcoming need to validate descriptors
up to different levels depending on why they were being read.

This requires tracking the validation level in a more granular fashion
in the Collection's caches. This in turn required dissociating the
uncommitted descriptors from the stored descriptors again, to allow the
stored descriptors cache to more closely mirror what's physically being
stored in the system.descriptor table.

Validation is now performed lazily and all descriptor queries now share
a common structure, be it by-ID lookups, by-name lookups, or full table
scans: cache lookups, reads, validation and hydration all happen in the
same sequence.

Despite significant rewriting of the inner workings of the descriptors
collection, this commit should not functionally change anything.

Informs #85263.

Release justification: low-risk, high-benefit change.
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86931)
<!-- Reviewable:end -->

Commit:
backupinfo: fix bug in writing names to metadata

Recent changes to the descriptors collection have uncovered some
brittleness in the logic for writing namespace entries for dropped
descriptors when creating backups.

This commit makes it:
- write deletion entries for descriptors other than tables, preventing
  possible catalog corruption.
- throw a better error message when attempting to write two identical
  keys consecutively.

Release justification: low-risk change
Release note: None

postamar commented:
Closed in favour of https://github.com/cockroachdb/cockroach/pull/87067 which is just better.
